make DebugMode wait on collectives before hashing outputs

bdhirsh opened the pull request:
this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169295
* #169023
* #171353

bdhirsh committed:
make DebugMode wait on collectives before hashing outputs

[ghstack-poisoned]

pianpwk reviewed (approved):
Thanks Brian!

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

yushangdi commented:
fyi I landed https://github.com/pytorch/pytorch/pull/169027!

yushangdi commented on torch/utils/_debug_mode.py:
```diff
@@ -967,6 +976,30 @@ def _dispatch_hash_hook(func, types, args, kwargs, result):
+                    if (
+                        a.name == "async_op"
+                        and type(a.type) is torch.BoolType
+                        and i < len(args)
```
I'm doing some testing with `_allgather_base_`, and I think this check might be too restrictive, because in eager, aync_op is default to True, and it doesn't show up in args, but we should wait for it

yushangdi replied:
This worked for me:

```
            for i, a in enumerate(func._schema.arguments):
                # Annoying, async_op is not a kwarg in the non-functional collectives.
                # I'm using OpOverload._schema to figure out if it exists
                if a.name == "async_op" and type(a.type) is torch.BoolType:
                    # Determine async_op value from args, kwargs, or default
                    if i < len(args):
                        async_op_value = args[i]
                    elif "async_op" in kwargs:
                        async_op_value = kwargs["async_op"]
                    elif a.has_default_value():
                        async_op_value = a.default_value
                    else:
                        async_op_value = False

                    if async_op_value:
                        for item in result:
                            if isinstance(item, torch.ScriptObject) and hasattr(
                                item, "wait"
                            ):
                                item.wait()
                    break

```

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh committed:
Update on "make DebugMode wait on collectives before hashing outputs"

this is an attempt to re-land https://github.com/pytorch/pytorch/pull/168119?brid=_FrxOZfmdeVjSSYD2a69yQ with a few tweaks:

(1) for non-functional collectives, only wait on the work item with `async=True`. [See comment](https://github.com/pytorch/pytorch/pull/168119#issuecomment-3582904056)

(2) For functional collectives, we can always call `wait_tensor` on the output.

The test in this PR will probably conflict with the test in https://github.com/pytorch/pytorch/pull/169027, so ill wait for that PR to land first and rebase.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -i
